File: scala-java/src/main/java/at/twinformatics/scalademo/java/ExceptionHandling.java
package at.twinformatics.scalademo.java;

import java.io.IOException;
import java.nio.file.Path;
import java.util.Optional;

import static java.nio.file.Files.readAllBytes;

/**
 * Simple exception handling example in Java. The checked `IOException`, thrown
 * by the `readAllBytes` method, must be handled.
 */
public class ExceptionHandling {

    Optional<String> readText(Path file) {
        try {
            return Optional.of(new String(readAllBytes(file)));
        } catch (IOException e) {
            return Optional.empty();
        }
    }

}

File: scala-java/src/main/scala/at/twinformatics/scalademo/scala/ExceptionHandling.scala
package at.twinformatics.scalademo.scala

import java.io.IOException
import java.nio.file.Path
import java.nio.file.Files.readAllBytes

/**
  * This example shows the advanced possibilities of Scala exception handling
  * using `PartialFunction`s as exception handler.
  */
object ExceptionHandling {

    def readText1(file: Path): Option[String] = {
        try {
            Some(new String(readAllBytes(file)))
        } catch {
            case e: IOException
                if e.getMessage.contains("Fatal") => throw e
            case e: IOException => None
        }
    }

    def readText2(file: Path): Option[String] = {
        try {
            Some(new String(readAllBytes(file)))
        } catch {
            handler(false)
        }
    }

    def handler(ignoreNull: Boolean):
    PartialFunction[Throwable, Option[String]] = {
        case e: NullPointerException if ignoreNull => None
        case _: IOException => None
    }

}
